include/deal.II/matrix_free/matrix_free.h

```diff
@@ -1668,6 +1668,9 @@ class MatrixFree : public Subscriptor
    * in general. The face range in @p loop runs from zero to
    * n_inner_face_batches() (exclusive), so this is the appropriate size if
    * you want to store arrays of data for all interior faces to be worked on.
+   * Note that it returns 0 unless either mapping_update_flags_boundary_faces or
+   * mapping_update_flags_inner_faces is set to a value different from
```

Comment:
Is this correct? I would have guessed that n_inner_face_batches()==0 unless n_inner_face_batches is set, while n_boundary_face_batches()==0 unless mapping_update_flags_boundary_faces is set.

Reply:
It is true for what we currently implement. We ought to implement one or the other one day, but we did not get around to do that yet as it's not lacking any functionality but just wastes memory and operations. Let me open an issue to track it.

Reply:
See #12763.

Reply:
Makes sense. @zjiaqi2018 if you want, you can take a quick look to see if you can implement the improvement.

Reply:
As a first attempt, you could try to just put an `if` around the block https://github.com/dealii/dealii/blob/7439f2b0b4d210d237e8df53f12f47c242cfac3f/include/deal.II/matrix_free/matrix_free.templates.h#L1800-L1805 as well as https://github.com/dealii/dealii/blob/7439f2b0b4d210d237e8df53f12f47c242cfac3f/include/deal.II/matrix_free/matrix_free.templates.h#L1818-L1835 in case the `inner` faces are not requested, and see if all matrix-free tests go through. This is not a complete solution because the dof indices will already include the indirections from neighboring inner faces, but it should help us with sorting out possible glitches. If that works, we could then remove the face-related stuff inside this function:
https://github.com/dealii/dealii/blob/7439f2b0b4d210d237e8df53f12f47c242cfac3f/include/deal.II/matrix_free/matrix_free.templates.h#L1040-L1044
If you want to check the first case, I can take over or at least help with the second more challenging case.